### PR TITLE
stack recovery for dup & ldelem.ref

### PR DESCRIPTION
**Pull request by 0xMcX**
Stack is not correctly discovered when multiple dup are used or when there are branches between.
For example : `ldarg dup .... dup ` or `ldarg brtrue ... dup`

ldelem.ref should be used instead of ldelema when not value type in Agile_NET.

### CODE REVIEW

**Comment by CreateAndInject** [comment]
I think `MethodStack` is a simple implementation which doesn't support any branches.
An instruction can have more than one source, so it's hard to decide pick which one as its previous instruction.
I think you should modify `Agile` code rather than `MethodStack` unless you can support branches completely.

**Comment by 0xMcX** [comment]
Code I sent you picks first branch, which is better than no branch at all and returning `null` or `dup` in stack. And stack recovery is not specific to `Agile`.
Branches support will be better, but there's always better ...

File: de4dot.code/deobfuscators/MethodStack.cs [hunk 1]
```diff
@@ -91,7 +79,8 @@ public static class MethodStack {
 					break;
 				if (instr.OpCode.Code == Code.Dup) {
 					pushes = 1;
-					pops = 0;
+					pops = 0;
+					instr = GetPushedArgInstructions(instructions, index, 1).GetEnd(0) ?? instr;
```
Comment (0xd4d) [inline de4dot.code/deobfuscators/MethodStack.cs:83]: This code seems to cause an infinite loop. Try deobfuscating .NET Reactor 4.7.0.0

**Comment by CreateAndInject** [comment]
```
				default:
					instr = instructions[instrIndex + 1];
					var flowInstr = instructions.FirstOrDefault(i => i.Operand is Instruction instruction && instruction == instr);
					if (flowInstr == null) return null;
					instrIndex = instructions.IndexOf(flowInstr);
					return flowInstr;
				}
```

If the `default` is reached, the `instr` is usually a conditional branch (br.true/bgt/ble....), it can fall through when the conditional isn't satisfied, so why give up it and pick a branch whose target is this instruction?

File: de4dot.code/deobfuscators/MethodStack.cs (hunk 1 above)
Comment (0xMcX) [reply de4dot.code/deobfuscators/MethodStack.cs]: Corrected but have no file to test

**Comment by 0xMcX** [comment]
Now default way should be fall through.

**Comment by CreateAndInject** [comment]
Can you attach the obfuscated file due to this PR?

**Comment by 0xMcX** [comment]
File sent by email. 
Added backtracking, should explore all branches without loops. 
Branch map is created only if needed, and cached based on the instructions list.

**Comment by 0xd4d** [comment]
Which one did you email?

I'll test this later this week.

**Comment by CreateAndInject** [comment]
I'm afraid whether there's some side effect? Even if the target obfusacted file work fine, is it possible to make some normal file not work, like what you find in .Net Reactor 4.7?

**Comment by 0xd4d** [comment]
The .NET Reactor bug was an infinite loop, I assume he's fixed it, I'll find out when I test it against some random files later this week.

**Comment by 0xd4d** [comment]
I just noticed you've changed the line endings in every file. Please revert that.

**Review by ghost — CHANGES_REQUESTED** [review]
.

**Comment by 0xMcX** [comment]
Indeed. Corrected to LF.

**Comment by 0xd4d** [comment]
It now throws an exception in a .NET Reactor obfuscated binary

```
ERROR: ------------------------------------------------------------------------------
ERROR: Stack trace:
ERROR:    at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)
ERROR:    at System.Collections.Generic.List`1.get_Item(Int32 index)
ERROR:    at de4dot.code.deobfuscators.MethodStack.UpdateState(IList`1 instructions, State state, PushedArgs pushedArgs)
ERROR:    at de4dot.code.deobfuscators.MethodStack.GetPushedArgInstructions(IList`1 instructions, Int32 index, Int32 numArgs)
ERROR:    at de4dot.code.deobfuscators.MethodStack.GetPushedArgInstructions(IList`1 instructions, Int32 index)
ERROR:    at de4dot.code.deobfuscators.TypesRestorerBase.DeobfuscateFields()
ERROR:    at de4dot.code.deobfuscators.TypesRestorerBase.DeobfuscateLoop()
ERROR:    at de4dot.code.deobfuscators.TypesRestorerBase.Deobfuscate()
ERROR:    at de4dot.code.deobfuscators.dotNET_Reactor.v4.Deobfuscator.DeobfuscateEnd()
ERROR:    at de4dot.code.ObfuscatedFile.Deobfuscate()
ERROR:    at de4dot.cui.FilesDeobfuscator.DeobfuscateAllFiles(IEnumerable`1 allFiles)
ERROR:    at de4dot.cui.FilesDeobfuscator.DeobfuscateAll()
ERROR:    at de4dot.cui.FilesDeobfuscator.DoIt()
ERROR:    at de4dot.cui.Program.Main(String[] args)
```

**Comment by 0xMcX** [comment]
Can you send me the test file ?

**Review by ghost — APPROVED** [review]
It worked now, thansk for the PR!

**Comment by 0xMcX** [comment]
You're welcome!

**Pull request merged**